addons/website_slides/models/slide_channel.py

```diff
@@ -748,6 +910,39 @@ def _action_request_access(self, partner):
                 )
         return activities
 
+    def _get_channel_invitation_link(self, partner_id=False):
+        ''' This will return the url used as:
+            - (1) course sharing link displayed in the invitation wizard.
+            - (2) hyperlink in the channel invitation email in template mail_notification_channel_invite.
+
+        Here are the different cases:
+        (1) : always return to the default /invite route
+        (2) - if partner_id has no user :
+                - if free sign up is allowed and the course has public visibility, then the user can browse
+                  freely the course and sign up only if desired. Therefore add auth_signup_token in url, kept
+                  in session and linking partner_id to created user on sign up.
+                - if no free sign up is allowed or the course is visible to members only, then the partner must
+                  directly receive a signup link, redirecting to the /invite route
+            - if partner_id has a user : send to the /invite route, with auth_login in url, so that if not
+                                         logged, login page will be prefilled with its login.
+
+        :param partner_id: (optional) partner id of the invited partner, in direct invitation emails.
+        :return: relevant channel invitation url, see above.
+        '''
+        url = '%s/slides/%s/invite?' % (self.get_base_url(), self.id)
+        if partner_id:
+            if partner_id.partner_share:
+                signup_allowed = self.env['res.users'].sudo()._get_signup_invitation_scope() == 'b2c'
+                partner_id.sudo().signup_prepare()
+                if self.visibility == 'members' or not signup_allowed:
+                    signup_url = partner_id._get_signup_url_for_action()[partner_id.id]
+                    url = '%s%s' % (signup_url, '&redirect=/slides/%s/invite?' % self.id)
```

Comment:
How does this work exactly?
When singup is not allowed, you can still send the partner a specific signup url?

Reply:
Yes, exactly. After a bit of search, I saw that the parameter returned by _get_signup_invitation_scope (called 'auth_signup_uninvited') does not appear in the direct invitation email process. Therefore, it is always possible to send a direct signup link to a partner. 

That is exactly what is done when granting portal access in portal module. signup_prepare() creates a token if necessary for the partner, and _get_signup_url_for_action() gets the appropriate url.

However, I see in auth_signup that there is a compute field 'signup_url' on partners that does the same, but also checks the write access rights to avoid writing on partners having already an internal user, if not allowed. Therefore, I switched to use partner_id.signup_url.